### PR TITLE
fix: harden the MCP↔Studio bridge against concurrency disconnects (v2.7.1)

### PR DESCRIPTION
**Pull request by garybenade**
## Problem

The HTTP poll-bridge between the MCP server and the Studio plugin dropped connections frequently under real agent usage — worst with parallel tool calls and multiple simultaneous agents (mid-session disconnects/timeouts).

## Root causes & fixes

**1. Double-dispatch race (core).** `BridgeService.getPendingRequest()` returned the oldest pending request but never marked it claimed, while the plugin polls every 0.5s. Any request slower than the poll interval was handed out again on the next poll (and the next…) — duplicate mutations plus a storm of `/response` posts. Requests now carry an in-flight `dispatchedAt`; one is only re-offered after a 10s redispatch TTL (lost-response recovery).

**2. Head-of-line starvation (core).** The same "return oldest, never claim" logic meant parallel/multi-agent requests starved behind the oldest in-flight one. Polls now serve the oldest *undispatched* request.

**3. Plugin requestId idempotency (plugin).** The plugin dedups by `requestId`: while `processing` it ignores repeats; once `done` it re-sends the *cached* response. This makes the server redispatch TTL safe — a re-dispatch can never re-execute side effects.

**4. Multi-agent port fragmentation (core).** Extra agents walked to their own ports (58742+) that the plugin never polls, so their calls dead-ended in 30s timeouts. Primary now binds the base port only; if it's taken, the agent enters proxy mode and forwards to the primary via the existing `ProxyBridgeService`, so every agent shares the one plugin connection. **This is the one behavioral change worth a careful look.**

**5. Silent response drops (plugin).** `sendResponse` swallowed failures in a bare `pcall` — a dropped post became an invisible 30s timeout. It now retries 3× and `warn()`s to Output.

**6. Adaptive polling (plugin).** 0.5s while work flows, backing off to 2s when idle, to keep HTTP volume well under Roblox's rate limit. I deliberately chose this over a server-held long-poll: `RequestAsync`'s default timeout is undocumented and its `Timeout` field can only *shorten* requests, so holding `/poll` open risks turning every hold into a plugin-side failure. Happy to revisit if you'd prefer long-poll.

## Also included

- Regenerated `MCPPlugin.rbxmx` / `MCPInspectorPlugin.rbxmx` — large diff, but tracked build output that must match the source changes.
- Version bump to **2.7.1** across packages + README.
- Repaired the broken `.eslintrc.json` (`@typescript-eslint/recommended` → `plugin:@typescript-eslint/recommended`) and the two pre-existing lint errors it surfaced, so `npm run lint` passes.

## Validation

- New `BridgeService` unit tests cover in-flight dispatch, no-starvation, and TTL re-dispatch (written test-first).
- Full suite **39/39 green**, typecheck clean, lint clean.
- **Validated in live Studio use** — a real multi-call agent session ran with zero disconnects/timeouts (vs. frequent drops before).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Stabilizes the MCP↔Studio HTTP bridge under parallel tool calls and multi‑agent runs by making dispatch idempotent and fair, and by funneling all agents through one plugin connection. Tightens port‑in‑use detection, hardens plugin dedupe, adds adaptive polling, and ships v2.7.1.

- **Bug Fixes**
  - Core dispatch: track in‑flight with a 10s redispatch TTL; serve the oldest undispatched request, only then TTL‑expired ones (prevents duplicates and starvation).
  - Plugin: requestId dedupe with a bounded cache that evicts only completed entries; resend cached result on repeats; retry `/response` up to 3× and warn.
  - Server: bind the base port only; fall back to proxy mode only when `isPortInUseError` is true (EADDRINUSE or the aggregate “All ports … are in use”); `listenWithRetry` exhaustion sets `EADDRINUSE`.
  - Adaptive polling: 0.5s while active, backing off to 2s when idle.
  - Tests/misc: add `BridgeService` and `isPortInUseError` tests; uptime check made non‑negative.

- **Dependencies**
  - Bump to `2.7.1` for `robloxstudio-mcp`, `robloxstudio-mcp-inspector`, `@robloxstudio-mcp/core`; regenerate `MCPPlugin.rbxmx` and `MCPInspectorPlugin.rbxmx`; fix ESLint extends to `plugin:@typescript-eslint/recommended`; update README version.

<sup>Written for commit 37c65ec20aab2a025fbe1fc23a596f4691c85609. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/boshyxd/robloxstudio-mcp/pull/121?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Adaptive polling cadence per-connection (faster with work, backs off when idle).
  * In-flight request tracking with redispatch timeout.
  * Depth-first descendant traversal utility to improve search performance.

* **Bug Fixes**
  * Response delivery retries (up to 3 attempts) with backoff and warnings.
  * Request deduplication to avoid duplicate handling.
  * HTTP binding only falls back to proxy mode for genuine port-in-use errors.

* **Tests**
  * Added coverage for in-flight dispatch and port/uptime behaviors.

* **Chores**
  * Version bumped to 2.7.1 and README metadata updated.

* **Style**
  * ESLint extends identifier aligned with plugin naming.

### CODE REVIEW

File: studio-plugin/src/modules/Communication.ts [hunk 1]
```diff
@@ -107,14 +107,55 @@ function processRequest(request: RequestPayload): unknown {
...
+// executed twice. While "processing" we ignore repeats; once "done" we re-send
+// the cached response instead of re-running the handler.
+const handledRequests = new Map<string, HandledRequest>();
+const handledOrder: string[] = [];
+const MAX_HANDLED_REQUESTS = 128;
+
+function rememberHandledRequest(requestId: string, entry: HandledRequest) {
+	if (!handledRequests.has(requestId)) {
+		handledOrder.push(requestId);
+		if (handledOrder.size() > MAX_HANDLED_REQUESTS) {
```
[the remark this answers is not in the document]
Comment (garybenade) [reply studio-plugin/src/modules/Communication.ts]: Fixed in 430064b. `rememberHandledRequest` no longer evicts by raw FIFO — it skips `processing` entries and drops only the oldest `done` one, so a re-dispatched in-flight id can no longer be re-executed under a >128-request burst. If the whole window is in-flight it allows a brief overflow, which is preferable to a duplicate side effect.

File: packages/core/src/server.ts [hunk 2]
```diff
@@ -94,20 +94,25 @@ export class RobloxStudioMCPServer {
...
+      // Base port already owned by another agent — share its bridge via proxy
+      // mode instead of binding a second primary the plugin would never poll.
       console.error(`Could not bind primary HTTP server: ${(err as Error).message}`);
       bridgeMode = 'proxy';
       primaryApp = undefined;
       const proxyBridge = new ProxyBridgeService(`http://localhost:${basePort}`);
       this.bridge = proxyBridge;
       this.tools = new RobloxStudioTools(this.bridge);
-      console.error(`All ports ${basePort}-${basePort + 4} in use — entering proxy mode (forwarding to localhost:${basePort})`);
+      console.error(`Port ${basePort} already in use — entering proxy mode (forwarding to the primary on localhost:${basePort})`);
```
[the remark this answers is not in the document]
Comment (garybenade) [reply packages/core/src/server.ts:122]: Fixed in 430064b. Added `isPortInUseError()`: the catch now re-throws genuine bind failures (EACCES, bad host, etc.) and only falls back to proxy mode on a port-in-use condition. `listenWithRetry`'s exhaustion error now carries code EADDRINUSE, and the classifier is covered by new unit tests.

File: packages/core/src/bridge-service.ts [hunk 3]
```diff
@@ -110,16 +117,23 @@ export class BridgeService {
...
+      if (request.dispatchedAt !== undefined && now - request.dispatchedAt < this.redispatchTimeout) {
+        continue;
+      }
       if (!oldestRequest || request.timestamp < oldestRequest.timestamp) {
         oldestRequest = request;
       }
     }
 
     if (oldestRequest) {
+      oldestRequest.dispatchedAt = now;
```
[the remark this answers is not in the document]
Comment (garybenade) [reply packages/core/src/bridge-service.ts:143]: Fixed in 1054df3. getPendingRequest now uses two-tier selection — it serves the oldest undispatched request first and only falls back to the oldest TTL-expired one when there is no fresh work, so a stalled re-offer can no longer leapfrog newer undispatched requests. Added a unit test covering it.

File: studio-plugin/src/modules/Communication.ts [hunk 4]
```diff
@@ -107,14 +107,55 @@ function processRequest(request: RequestPayload): unknown {
...
+			const oldest = handledOrder[0];
+			handledOrder.remove(0);
+			handledRequests.delete(oldest);
+		}
```
[the remark this answers is not in the document]
Comment (garybenade) [reply studio-plugin/src/modules/Communication.ts:169]: The processing-eviction half is fixed in 430064b — eviction skips in-flight entries and drops only the oldest done one. I deliberately left out the done-entry TTL: the plugin gets at most one request per poll (~2/sec), so a 128-entry window takes ~64s to fill, meaning a done entry is never evicted until long past both the 10s redispatch window and the 30s request timeout — by then its response can no longer be re-requested. Evicting oldest-done-first also protects recently completed entries. Glad to add an explicit completedAt TTL if you want the guarantee in code.

File: studio-plugin/MCPInspectorPlugin.rbxmx [hunk 5]
```diff
@@ -138,19 +138,56 @@ local function processRequest(request)
...
+		table.insert(handledOrder, _requestId_1)
+		if #handledOrder > MAX_HANDLED_REQUESTS then
+			local oldest = handledOrder[1]
+			table.remove(handledOrder, 1)
+			handledRequests[oldest] = nil
+		end
+	end
+	local _requestId_1 = requestId
+	local _entry = entry
+	handledRequests[_requestId_1] = _entry
```
[the remark this answers is not in the document]
Comment (garybenade) [reply studio-plugin/MCPInspectorPlugin.rbxmx:182]: Same dedupe logic as the Communication.ts source (this is the compiled artifact). Fixed in 430064b — eviction skips processing entries and drops only the oldest done one. The done-entry TTL refinement I skipped on purpose: at ~2 requests/sec (one per poll) the 128-entry window takes ~64s to fill, well past the 10s redispatch window, so a done entry is never evicted while still re-dispatchable.

File: packages/core/src/http-server.ts [hunk 6]
```diff
@@ -396,6 +396,18 @@ export function createHttpServer(tools: RobloxStudioTools, bridge: BridgeService
...
+ * Classify a primary-bind failure. Returns true only when the base port is
+ * genuinely occupied, so falling back to proxy mode is the correct response.
+ * Other bind errors (EACCES, bad host, ...) must NOT be masked by proxy mode —
+ * they should surface instead of silently forwarding to a primary that isn't there.
+ */
+export function isPortInUseError(err: unknown): boolean {
+  if (!err) return false;
+  const e = err as { code?: string; message?: string };
+  return e.code === 'EADDRINUSE' || /in use/i.test(e.message ?? '');
+}
```
[the remark this answers is not in the document]
Comment (garybenade) [reply packages/core/src/http-server.ts:412]: Fixed in 37c65ec. Anchored the message fallback to /all ports .* are in use/i (the exact aggregate phrasing listenWithRetry emits), so an unrelated error whose message merely contains 'in use' no longer trips the proxy-mode fallback. The explicit code === 'EADDRINUSE' check still covers the real paths (raw bind error and the stamped aggregate). Added a unit test pinning that a generic 'in use' message classifies as false.
